### PR TITLE
samples: npmx: create basic module layout

### PR DESCRIPTION
**Pull request by wilu-nordic**

### CODE REVIEW

File: drivers/npmx/CMakelists.txt [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# Copyright (c) 2021 Nordic Semiconductor ASA
```
Comment (jakub-uC) [inline drivers/npmx/CMakelists.txt:1]: 2022

File: drivers/npmx/Kconfig [hunk 2]
```diff
@@ -0,0 +1,23 @@
+
+config NPMX_ENABLE
+	bool "Enabling npmx"
+    default n
```
Comment (jakub-uC) [inline drivers/npmx/Kconfig:4]: No need for this, `default` is `n` by default ;-)

File: drivers/npmx/npmx_config_npm1300.h [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2019, 2022 Nordic Semiconductor ASA
```
Comment (jakub-uC) [inline drivers/npmx/npmx_config_npm1300.h:2]: `Copyright (c) 2022 Nordic Semiconductor ASA` is enough

File: drivers/npmx/npmx_driver.c [hunk 4]
```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2019, 2022 Nordic Semiconductor ASA
```
Comment (jakub-uC) [inline drivers/npmx/npmx_driver.c:2]: `Copyright (c) 2022 Nordic Semiconductor ASA` is enough

File: drivers/npmx/npmx_driver.c [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+static void generic_callback(npmx_instance_t *pm, npmx_callback_type_t type, uint32_t arg)
+{
+	LOG_DBG("[%lld]\t %s:\t\t", k_uptime_get(), npmx_callback_to_str(type));
+	for (uint8_t i = 0; i < 8; i++) {
+		if ((1U << i) & arg) {
+			LOG_DBG("%s, ", npmx_callback_bit_to_str(type, i));
+		}
+	}
+	LOG_DBG("\n");
```
Comment (jakub-uC) [inline drivers/npmx/npmx_driver.c:32]: I think it is not needed.

File: drivers/npmx/npmx_driver.c [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+	return 0;
+};
+
+bool npmx_driver_register_cb(npmx_callback_t cb, npmx_callback_type_t type)
+{
+	if (p_npmx_instance && cb) {
+		p_npmx_instance->registered_cb[(uint32_t)type] = cb;
+		return true;
+	} else {
```
Comment (jakub-uC) [inline drivers/npmx/npmx_driver.c:78]: obsolete `else`

File: drivers/npmx/npmx_driver.c [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		return false;
+	}
+}
+
+bool npmx_driver_event_interrupt_enable(npmx_event_group_t event, uint8_t flags_mask)
+{
+	if (p_npmx_instance) {
+		return npmx_event_interrupt_enable(p_npmx_instance, event, flags_mask);
+	} else {
```
Comment (jakub-uC) [inline drivers/npmx/npmx_driver.c:87]: obsolete `else`

File: drivers/npmx/npmx_driver.c [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+		return false;
+	}
+}
+
+static const struct sensor_driver_api npmx_driver_api = {
+	.sample_fetch = NULL,
+	.channel_get = NULL,
+// #ifdef CONFIG_BH1749_TRIGGER
```
Comment (jakub-uC) [inline drivers/npmx/npmx_driver.c:95]: Do we need these comments here?

File: drivers/npmx/npmx_driver.h [hunk 9]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
```
Comment (jakub-uC) [inline drivers/npmx/npmx_driver.h:2]: `2022`

File: drivers/npmx/npmx_glue.c [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+			   uint16_t register_address,
+			   uint8_t *p_data,
+			   uint32_t num_of_bytes)
+{
+	const struct device *i2c_dev = (struct device *)p_inst->p_backend;
+	uint8_t wr_addr[2];
+	struct i2c_msg msgs[2];
+
+	/* npmx register address */
+	wr_addr[0] = (uint8_t)(register_address >> 8U) & 0xFF;
```
Comment (jakub-uC) [inline drivers/npmx/npmx_glue.c:20]: Please do not use magic numbers.

File: drivers/npmx/npmx_glue.c [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	msgs[0].buf = wr_addr;
+	msgs[0].len = 2U;
+	msgs[0].flags = I2C_MSG_WRITE;
+
+	/* Data to be written, and STOP after this. */
+	msgs[1].buf = p_data;
+	msgs[1].len = num_of_bytes;
+	msgs[1].flags = I2C_MSG_WRITE | I2C_MSG_STOP;
+
+	return i2c_transfer(i2c_dev, &msgs[0], 2, p_inst->address);
```
Comment (jakub-uC) [inline drivers/npmx/npmx_glue.c:35]: `i2c_transfer(i2c_dev, msgs, 2, p_inst->address);` -> this will be shorter

File: drivers/npmx/npmx_glue.h [hunk 12]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2019, 2022 Nordic Semiconductor ASA
```
Comment (jakub-uC) [inline drivers/npmx/npmx_glue.h:2]: `2022` is enough

File: drivers/npmx/npmx_glue.c [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+	msgs[0].buf = wr_addr;
+	msgs[0].len = 2U;
+	msgs[0].flags = I2C_MSG_WRITE;
+
+	/* Read from device. STOP after this. */
+	msgs[1].buf = p_data;
+	msgs[1].len = num_of_bytes;
+	msgs[1].flags = I2C_MSG_READ | I2C_MSG_STOP;
+
+	return i2c_transfer(i2c_dev, &msgs[0], 2, p_inst->address);
```
Comment (jakub-uC) [inline drivers/npmx/npmx_glue.c:63]: `return i2c_transfer(i2c_dev, msg, 2, p_inst->address);`

File: drivers/npmx/npmx_glue.h [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+ * @brief Macro for placing a runtime assertion.
+ *
+ * @param expression Expression to be evaluated.
+ */
+#define NPMX_ASSERT(expression)
+
+/**
+ * @brief Function for I2C write.
+ *
+ * @param inst Pointer to backend instance
```
Comment (jakub-uC) [inline drivers/npmx/npmx_glue.h:35]: Can we keep convention `param[in/out]` ?

File: dts/bindings/npmx/nordic,npm1300.yaml [hunk 15]
```diff
@@ -0,0 +1,17 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
```
Comment (jakub-uC) [inline dts/bindings/npmx/nordic,npm1300.yaml:2]: `2022`

File: drivers/npmx/npmx_trigger.c [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+	gpio_init_callback(&data->gpio_cb, npmx_gpio_callback, BIT(int_gpio->pin));
+
+	err = gpio_add_callback(int_gpio->port, &data->gpio_cb);
+	if (err < 0) {
+		LOG_ERR("Failed to set GPIO callback: %d", err);
+		return err;
+	}
+
+	err = gpio_pin_interrupt_configure_dt(&config->int_gpio, GPIO_INT_LEVEL_HIGH);
+	if (err < 0) {
```
Comment (jakub-uC) [inline drivers/npmx/npmx_trigger.c:70]: What about `err != 0` ?

File: drivers/npmx/npmx_driver.c (hunk 8 above)
Comment (wilu-nordic) [reply drivers/npmx/npmx_driver.c:95]: Removed, it is not used

File: drivers/npmx/npmx_config.h [hunk 17]
```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) 2019, 2022 Nordic Semiconductor ASA
```
Comment (jakub-uC) [inline drivers/npmx/npmx_config.h:2]: `2022`

File: drivers/npmx/npmx_driver.c [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+	p_npmx_instance = &data->npmx_instance;
+
+	p_npmx_instance->generic_cb = generic_callback;
+
+	if (npmx_core_init(&data->npmx_instance) == false) {
+		LOG_ERR("Unable to init npmx device");
+		return -EIO;
+	}
+
+	/* TODO: clear all events before enabling interrupts */
```
Comment (jakub-uC) [inline drivers/npmx/npmx_driver.c:63]: Is this comment relevant?

File: drivers/npmx/npmx_driver.h [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+struct npmx_config {
+	const struct i2c_dt_spec i2c;
+	const struct gpio_dt_spec int_gpio;
+};
+
+/**
+ * @brief Register callback handler to specified callback type event/error
+ * If no callback is registered, generic_cb is called.
+ * This function should be called in other drivers init function e.g. charger, vbus,
+ * with initialization priority higher then @ref CONFIG_NPMX_INIT_PRIORITY
```
Comment (magp-nordic) [inline drivers/npmx/npmx_driver.h:31]: I think there is a typo in word "then".

File: samples/npmx/charger_and_events/src/main.c [hunk 20]
```diff
@@ -6,443 +6,49 @@
...
 
-	npm1300_adc_ntc_set(&pm, NPM1300_BATTERY_NTC_TYPE_10_K);
+	/* Temporary tests: use GPIO 0 as interrupt output */
+	npmx_gpio_mode_set(npmx_instance, NPMX_GPIO_INSTANCE_0, NPMX_GPIO_MODE_OUTPUT_IRQ);
 
-	npm1300_check_errors(&pm);
+	/* Temporary tests: enable interrupt from vbusin events */
+	npmx_driver_event_interrupt_enable(NPMX_EVENT_GROUP_VBUSIN_VOLTAGE,
+				       NPMX_EVENT_GROUP_VBUSIN_DETECTED_MASK
+				       | NPMX_EVENT_GROUP_VBUSIN_REMOVED_MASK);
```
Comment (magp-nordic) [inline samples/npmx/charger_and_events/src/main.c:50]: I think or sign `|` should not be at the beginning of the line, but at the end of the previous one instead. Also, arguments should be aligned.

File: drivers/npmx/npmx_trigger.c [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+	npmx_instance_t *npmx_instance = &data->npmx_instance;
+
+	npmx_interrupt(npmx_instance);
+
+	npmx_proc(npmx_instance);
+
+	gpio_pin_interrupt_configure_dt(&config->int_gpio, GPIO_INT_LEVEL_HIGH);
+}
+
+int npmx_gpio_interrupt_init(const struct device *dev)
```
Comment (magp-nordic) [inline drivers/npmx/npmx_trigger.c:40]: Why this function is defined in separate file if it is declared in `npmx_driver.h`?

File: drivers/npmx/npmx_driver.c (hunk 18 above)
Comment (wilu-nordic) [reply drivers/npmx/npmx_driver.c:63]: Yes, it have to be implemented in [NRFX-2146](https://projecttools.nordicsemi.no/jira/browse/NRFX-2146)

File: drivers/npmx/npmx_config_npm1300.h [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DRIVERS_NPMX_NPMX_CONFIG_NPM1300_H__
+#define ZEPHYR_DRIVERS_NPMX_NPMX_CONFIG_NPM1300_H__
+
+#ifndef ZEPHYR_DRIVERS_NPMX_NPMX_CONFIG_H__
+#error "This file should not be included directly. Include nrfx_config.h instead."
+#endif
+
```
[the remark this answers is not in the document]
Comment (wilu-nordic) [reply drivers/npmx/npmx_config_npm1300.h:13]: npm1300_fp1 is selected in `npmx/mdk`

File: drivers/npmx/npmx_config.h [hunk 23]
```diff
@@ -0,0 +1,12 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DRIVERS_NPMX_NPMX_CONFIG_H__
+#define ZEPHYR_DRIVERS_NPMX_NPMX_CONFIG_H__
```
[the remark this answers is not in the document]
Comment (wilu-nordic) [reply drivers/npmx/npmx_config.h:8]: first `ZEPHYR_DRIVERS_NPMX` is from directories 
second `NPMX_CONFIG_H__` is from file name

File: drivers/npmx/npmx_driver.h [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+#include <npmx.h>
+#include <zephyr/types.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+
+struct npmx_data {
+	const struct device *dev;
+	npmx_instance_t npmx_instance;
+	struct k_work work;
+	struct gpio_callback gpio_cb;
```
[the remark this answers is not in the document]
Comment (wilu-nordic) [reply drivers/npmx/npmx_driver.h:19]: Here is Zephyr coding standard, not NRFX

File: drivers/npmx/npmx_trigger.c (hunk 21 above)
Comment (wilu-nordic) [reply drivers/npmx/npmx_trigger.c:40]: Zephyr coding standard, I did similar to BH1749 sensor

**Comment by wilu-nordic** [comment]
> Shouldn't folder name for examples be named `npm1300` instead of `npmx`? It has examples for specific version of npmx.

Difficult to resolve for now. For example all nPM devices should implement charger and events.

**Review by magp-nordic — APPROVED** [review]

**Pull request closed without merging**